### ISSUE TITLE
Forward port for external zone is reachable from internal zone and gets all port-related traffic

### ISSUE THREAD

**Issue by agross** [firewalld/firewalld#258]
Hi,

I have to admit that I'm a bit of a iptables noob, but I believe I get the gist of it.

I run a router using firewalld and have a dial-up connection that needs to provide some services (HTTP, FTP, etc.). Some services are running on the router itself (HTTP), some need to be redirected to other machines (FTP). I have no problems with the services running on the router, but those running on other machines.

My router has 172.16.0.3 (internal zone) and a PPP connection + ethernet in the external zone.

```sh
$ firewall-cmd --info-zone internal
internal (active)
  target: default
  icmp-block-inversion: no
  interfaces: ens192
  sources:
  services: ntp cockpit dhcpv6-client http mdns ssh https afp
  ports:
  protocols:
  masquerade: no
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:

$ firewall-cmd --permanent --info-zone external
external (active)
  target: default
  icmp-block-inversion: no
  interfaces: ppp0 ens224
  sources:
  services: https ssh http
  ports:
  protocols:
  masquerade: yes
  forward-ports: port=10021:proto=tcp:toport=10021:toaddr=172.16.0.10
        port=21:proto=tcp:toport=21:toaddr=172.16.0.10
  source-ports:
  icmp-blocks:
  rich rules:
```

My first impression was that a some port forwarding should suffice to publish FTP to the outside world:

```sh
 firewall-cmd --add-forward-port=port=21:proto=tcp:toport=21:toaddr=172.16.0.10 --zone=external
 firewall-cmd --add-forward-port=port=10021:proto=tcp:toport=10021:toaddr=172.16.0.10 --zone=external
```

This makes the FTP service available to the outside world but I found that when I connect to any FTP server on the internet from a machine behind the router (e.g. not the router itself), I instead reach my own FTP server.

My second attempt has been to create a FTP service and add that to the external zone (removing the mapping beforehand). But now my FTP server is not reachable from the outside world:

```sh
cat /etc/firewalld/services/syno-ftp.xml
<?xml version="1.0" encoding="utf-8"?>
<service>
  <short>Synology FTP Server</short>
  <description>Synology FTP Server on 172.16.0.10</description>
  <port port="21" protocol="tcp"/>
  <port port="10021" protocol="tcp"/>
  <module name="ftp"/>
  <destination ipv4="172.16.0.10"/>
</service>

$ firewall-cmd --zone=external --add-service=syno-ftp
success

$ iptables -L
...
Chain IN_external_allow (1 references)
target     prot opt source               destination
ACCEPT     tcp  --  anywhere             anywhere        tcp dpt:http ctstate NEW
ACCEPT     tcp  --  anywhere             172.16.0.10     tcp dpt:21 ctstate NEW
ACCEPT     tcp  --  anywhere             172.16.0.10     tcp dpt:10021 ctstate NEW

$ tcpdump -i ppp0 -n 'tcp port 21'
listening on ppp0, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
19:56:51.942657 IP 85.25.226.101.51740 > 91.41.223.49.ftp: Flags [S], seq 2162915807, win 8192, options [mss 1460,nop,wscale 8,nop,nop,sackOK], length 0
...
```

My question is: What is the correct way to configure FTP to be available externally, but not "eat up" my outgoing FTP connections.

**Comment by wlnx** [comment]
I definitely [confirm the issue](https://www.centos.org/forums/viewtopic.php?f=50&t=65044).

**Comment by guard43ru** [comment]
I confirm that, when setting up forward rules in one zone, they somehow work in other zones.

**Comment by erig0** [comment]
@nils-ballmann, Thanks for the thorough explanation. I took a quick look at the iptables rules generated by firewalld, but didn't see anything obviously wrong.

If you disable the masquerade on the external zone, are you still redirected to server3 (from server1)?

**Comment by nils-ballmann** [comment]
I can't do this, because it's my production environment. I'll try to test this with a lab setup or even a pure VM setup.
I will have the first results probably tomorrow noon (CEST).

Expected behavior, in case the masquerading is turned off, is to not be able to reach `host1` at all.
Buggy behavior should be to still reach `server3`, right?

Btw I'm working on RHEL7:

    host # yum list installed firewalld
    firewalld.noarch                         0.4.4.4-6.el7                           @repo1/7.3

Are the port forwardings not disabled when the masquerading is disabled? I *think* I read sth like this.

**Comment by erig0** [comment]
> Expected behavior, in case the masquerading is turned off, is to not be able to reach host1 at all.
Buggy behavior should be to still reach server3, right?

Correct. I'm wondering if there is an interaction between masquerade and forward-ports that's causing this unexpected behavior.

**Comment by nils-ballmann** [comment]
Could not reproduce the issue in a pure VBox Setup with a CentOS7.4 VM.

Going for the lab-setup tomorrow.

**Comment by emmettculley** [comment]
I see the same issue.  I set up a port foward to forward port 80 on th public zone to an IP address on the trusted (LAN) zone.  From then on, anytime I attempt to go to any website outside (on the internet) I am redirected to that local webserver that the port forward to configured for.

It is redirecting ALL accesses to port 80, internal and external to the local IP address.  Not very useful.

Is there a way to tell a port forward to only look at traffic comming into the public zone from outside?

I am using the firewalld GUI app.

**Comment by fossilet** [comment]
Rich rule does not work either.

```
[root@router network-scripts]# firewall-cmd --list-all
external (active)
  target: default
  icmp-block-inversion: no
  interfaces: ens192
  sources: 
  services: ssh http https
  ports: 
  protocols: 
  masquerade: yes
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
	rule family="ipv4" forward-port port="80" protocol="tcp" to-port="80" to-addr="10.0.0.13"
```

When I connect to outside website, it is refused as the port forwarding rule.

**Comment by fossilet** [comment]
Using iptables-service for now temporarily , but still waiting for this to be fixed.

**Comment by tudorbarascu** [comment]
@agross I had the same problem. I solved it like this:
I have an external zone with an EXTERNAL_IP + a work zone with a network like``192.168.222.0/24``.

To allow port 587 forward from external zone it was simple:
``firewall-cmd --zone=external --add-forward-port=port=587:proto=tcp:toport=587:toaddr=192.168.222.4``

Adding the same for my work zone was working if I was connecting to my mail server (on the same network but accessing it from the EXTERNAL_IP but as soon I tried to connect to outside servers (port 587) everything was actually forwarded to the internal mail server.

So I added this rich rule and it worked:
``firewall-cmd --permanent --zone=work --add-rich-rule='rule family=ipv4 source address=192.168.222.0/24 destination address=EXTERNAL_IP/32 forward-port port=587 protocol=tcp to-port=587 to-addr=192.168.222.4'``

Notice the ``destination address=EXTERNAL_IP/32`` part where you should put your EXTERNAL_IP.

Basically, I'm telling firewalld that he should forward that port ONLY for connections with the destination (EXTERNAL_IP) that come from the internal network.
You should do the same and it will work.

@nils-ballmann @emmettculley  This should also work for you.

@fossilet When you're connecting to outside.. you're not using the external zone.. but the internal one (or how it's called on your system) and you haven't posted the firewall rules for that. This should also work for you.

**Comment by agross** [comment]
> Notice the destination address=EXTERNAL_IP/32 part where you should put your EXTERNAL_IP.

Thank you @tudorbarascu for putting life back into this old issue. Unfortunately, my external IP is dynamic and I believe it would complicate things for me if I needed to add a new (and remove the old) rule with `destination address=EXTERNAL_IP/32` every time I am forced to re-establish my PPP connection.

**Comment by zircop** [comment]
@erig0 Have you come any closer to solving this issue?

My setup:

```
firewall-cmd --get-default-zone
FedoraServer
```
```
FedoraServer
  target: default
  icmp-block-inversion: no
  interfaces: eno2
  sources: 
  services: ssh dhcpv6-client cockpit dns https http imap smtp smtps imaps
  ports: 5000/tcp 5001/tcp 2812/tcp 465/tcp 4352/tcp
  protocols: 
  masquerade: no
  forward-ports: port=53:proto=udp:toport=5300:toaddr=
	port=53:proto=tcp:toport=5300:toaddr=
  source-ports: 
  icmp-blocks: 
  rich rules:
```
```
trusted (active)
  target: ACCEPT
  icmp-block-inversion: no
  interfaces: 
  sources: 192.168.0.0/16
  services: 
  ports: 
  protocols: 
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
```
So basically, DNS lookups from 192.168.X.X should go to port 53 and DNS lookups from anywhere else should go to port 5300, all DNS lookups except from loopback interface goes to 5300.

I have tried with eno2 bound to FedoraServer and unbound, both with same result.

Calling any service on the server not listed in any rules gets blocked from outside 192.168.X.X scope so the trusted zone assignment appears to be working.

**Comment by erig0** [comment]
> @erig0 Have you come any closer to solving this issue?

How forward ports is implemented was recently changed in 7235f76a818329e3aaba2f8c1444f15b4af586c2. It's quite possible the rework also fixed this issue as a side effect. I will test it on my side, but if anyone could verify this bug against master that would be helpful.

**Comment by zircop** [comment]
I have built master branch (0.6.999), installed rpms on an internal half dev half production server and tried two scenarios:

```
(Server IP 192.168.20.50)

internal (active)
  target: default
  icmp-block-inversion: no
  interfaces: 
  sources: 192.168.20.1
  services: dhcpv6-client dns mdns samba-client ssh
  ports: 
  protocols: 
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 

FedoraServer (active)
  target: default
  icmp-block-inversion: no
  interfaces: eno2
  sources: 
  services: cockpit dhcpv6-client dns http https ssh
  ports: 8000/tcp 5010/tcp
  protocols: 
  masquerade: no
  forward-ports: port=53:proto=udp:toport=5300:toaddr=
  source-ports: 
  icmp-blocks: 
  rich rules: 
```
**Result**: DNS queries from both 192.168.20.1 and 192.168.20.147 gets forwarded to 5300.


```
(Server IP 192.168.20.50)

internal (active)
  target: default
  icmp-block-inversion: no
  interfaces: 
  sources: 192.168.20.1
  services: dhcpv6-client dns mdns samba-client ssh
  ports: 8000/tcp 5010/tcp
  protocols: 
  masquerade: no
  forward-ports: port=53:proto=udp:toport=5300:toaddr=
  source-ports: 
  icmp-blocks: 
  rich rules: 

FedoraServer (active)
  target: default
  icmp-block-inversion: no
  interfaces: eno2
  sources: 
  services: cockpit dhcpv6-client dns http https ssh
  ports: 
  protocols: 
  masquerade: no
  forward-ports:
  source-ports: 
  icmp-blocks: 
  rich rules: 
```
**Result**: DNS queries from 192.168.20.1 gets forwarded to 5300 while DNS queries from 192.168.20.147 goes to 53.

I’m not able to test this on my other server with the desired setup since it’s a production server, but I think this test is close enough to verify that at least a part of the issue still exists.

Do you want any iptables dumps or anything else that can help you find a solution? If so, tell me what commands you need me to run.

If it helps, I can leave the server with a master branch firewalld setup for a while and update it to latest master after changes to rerun the test. However, in June I need to put it in production and I’m not sure I’m allowed to keep experimenting with firewalld on it then.

**Comment by nils-ballmann** [comment]
I plan to use [GNS3](https://www.gns3.com/) to simulate the issue with a CentOS VM as masquerading router at work. A colleague of mine has configured a base-simulation already, so it probably won't be much work. Especially if you use vagrant VMs. 

But I got so much to do otherwise, that I don't find time to do it. Not even in a HW setup.

The last time I tried to do this with VMs, I tried to build the network myself and failed. But with GNS3 this should be way easier. Maybe someone of you wants to give it a shot?

**Comment by erig0** [comment]
I have found this occurs under the following conditions:
- the zone with port forwarding is also the **default zone** (e.g. public, external)
- a second zone (e.g. internal, trusted) uses a **--set-target** other than *default*

Why does it occur?
- in the mangle table dispatch to the secondary zone is done using a "jump". This allows the packets to also drift to the default zone and therefore port forwarding occurs. This should not happen. Zone drifting is a bug and needs to be addressed, see #441.

Workaround 1:
- Use a dummy zone as the default. A good choice would be "drop" or "reject".

Workaround 2:
- Use the --set-target=default in the secondary zone. This has other implications, so the above will probably work better.

**Comment by agross** [comment]
@erig0 Thank you for investigating. For the conditions you listed, do both need to be satisfied or just one?

**Comment by erig0** [comment]
> @erig0 Thank you for investigating. For the conditions you listed, do both need to be satisfied or just one?

Not exactly. It can occur other ways, but it all boils down to zone drifting. Another example:
- zoneA is source based (chain _ZONES_SOURCE occurs before _ZONES)
- zoneB is interface based and has port forwarding enabled

**Comment by tudorbarascu** [comment]
@erig0 I'm using Centos 8 with firewalld 0.7.0 and the issue is still present. Should I wait for another version or should I give more details here?

**Comment by erig0** [comment]
@tudorbarascu, wait for the next release. The fix will be in all the next stable releases which will occur around April 1st.

**Comment by tudorbarascu** [comment]
@erig0 Will test and get back here in April just to confirm. Thanks a lot!

**Comment by tudorbarascu** [comment]
@erig0 Hello. Tested again with Centos 8 and firewalld 0.8.0, still the same behavior (tried with zone drifting on/off, changing default zone, everything that I can think off).

**Comment by erig0** [comment]
@tudorbarascu, I had to re-read your [report above](https://github.com/firewalld/firewalld/issues/258#issuecomment-453733565). It sounds like the firewalld node is a gateway/router and you're adding a `forward-port` to the `work` zone. Then other hosts/machines in the `work` zone are trying to connect to dest port 587. The traffic from `work` --> internet in redirected to the host specified in your `forward-port`. This is 100% expected. The traffic ingresses the `work` zone which is subject to the `forward-port` you configured. The only way to prevent this is to scope the `forward-port` with a rich rule as you've done above.

**Comment by tudorbarascu** [comment]
@erig0 Thanks for responding.

Indeed, the firewalld node is a gateway/router and from the EXTERNAL zone (with the public IP) and I have a port forward to a LAN ip.

I simply expected that a LAN client can make a working request to that external IP/port.

"The traffic from work --> internet in redirected to the host specified in your forward-port". Well.. that's just it, it doesn't.

**Comment by erig0** [comment]
> "The traffic from work --> internet in redirected to the host specified in your forward-port". Well.. that's just it, it doesn't.

This contradicts what you said above..

> Adding the same for my internal zone (work) was working if I was trying to connect ONLY to my mail server.
> If I tried to access other mail servers on port 587 I was forwarded to my mail server.

**Comment by tudorbarascu** [comment]
@erig0 Indeed. Sorry. I meant to say that on a default external interface on a router/gateway, a port forward works only for external connections, not from internal ones. The problem outlined at first (with the mail) had my workaround.

Thank you.

**Comment by tudorbarascu** [comment]
I just realised that the workaround works only when masquerading is enabled on the work zone. That's why I couldn't replicate anymore (as I needed to disable masquerading in the LAN).
